phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -555,6 +560,24 @@ private void openConnection() throws SQLException {
         }
     }
 
+    /*
+    We create a long-lived hbase connection to run invalidate cache RPCs. We override
+    CUSTOM_CONTROLLER_CONF_KEY to instantiate InvalidateMetadataCacheController which has
+    a special priority for invalidate metadata cache operations.
+ */
+    public Connection getInvalidateMetadataCacheConnection() throws IOException {
+        if (invalidateMetadataCacheConnection != null) {
+            return invalidateMetadataCacheConnection;
+        }
+        // TODO Do we need a double locking here?
+        Configuration clonedConfiguration = PropertiesUtil.cloneConfig(this.config);
+        clonedConfiguration.setClass(CUSTOM_CONTROLLER_CONF_KEY,
+                InvalidateMetadataCacheControllerFactory.class, RpcControllerFactory.class);
+        invalidateMetadataCacheConnection = ConnectionFactory.createConnection(clonedConfiguration);
```

Comment:
`openConnection()` method uses `HBaseFactoryProvider.getHConnectionFactory()`

Reply:
Refactored openConnection and closeConnection methods [here](https://github.com/apache/phoenix/pull/1748/commits/4a2138f8124fc1d09a66bd1e6f43f55c518595de).